addons/purchase_stock/models/stock_rule.py

```diff
@@ -122,12 +133,12 @@ def _run_buy(self, procurements):
             procurements = self._merge_procurements(procurements_to_merge)
 
             po_lines_by_product = {}
-            grouped_po_lines = groupby(po.order_line.filtered(lambda l: not l.display_type and l.product_uom == l.product_id.uom_po_id), key=lambda l: l.product_id.id)
-            for product, po_lines in grouped_po_lines:
-                po_lines_by_product[product] = self.env['purchase.order.line'].concat(*po_lines)
+            grouped_po_lines = groupby(po.order_line.filtered(lambda l: not l.display_type and l.product_uom == l.product_id.uom_po_id), key=self._get_po_line_grouping_key)
```

Comment:
You still need to use `lambda` to send only 1 line at a time in the function
```suggestion
            grouped_po_lines = groupby(po.order_line.filtered(lambda l: not l.display_type and l.product_uom == l.product_id.uom_po_id), key=lambda l: self._get_po_line_grouping_key(l))
```

Reply:
nevermind, that works